src/js/core/game_state.ts

```diff
@@ -1,65 +1,54 @@
-/* typehints:start */
-import { Application } from "../application";
-import { StateManager } from "./state_manager";
-/* typehints:end */
-
-import { globalConfig } from "./config";
+import { MUSIC } from "@/platform/sound";
+import type { Application } from "../application";
 import { ClickDetector } from "./click_detector";
-import { logSection, createLogger } from "./logging";
+import { globalConfig } from "./config";
 import { InputReceiver } from "./input_receiver";
-import { waitNextFrame } from "./utils";
+import { createLogger, logSection } from "./logging";
 import { RequestChannel } from "./request_channel";
-import { MUSIC } from "../platform/sound";
+import type { StateManager } from "./state_manager";
+import { waitNextFrame } from "./utils";
 
 const logger = createLogger("game_state");
 
 /**
  * Basic state of the game state machine. This is the base of the whole game
  */
 export class GameState {
-    /**
-     * Constructs a new state with the given id
-     * @param {string} key The id of the state. We use ids to refer to states because otherwise we get
-     *                     circular references
-     */
-    constructor(key) {
-        this.key = key;
+    public app: Application = null;
+    public readonly key: string;
+    public inputReceiver: InputReceiver;
 
-        /** @type {StateManager} */
-        this.stateManager = null;
+    /** A channel we can use to perform async ops */
+    protected asyncChannel = new RequestChannel();
+    protected clickDetectors: ClickDetector[] = [];
 
-        /** @type {Application} */
-        this.app = null;
+    /** @todo review this */
+    protected htmlElement: HTMLElement | undefined;
```

Comment:
review what specifically

Reply:
The main state elements could be exposed better, for example via getter that throws depending on lifecycle state (e.g. accessing it before onEnter). If performance is critical, a TypeScript decorator can be used to make a "lazy" getter that overwrites itself with the computed value.